README.md
# CheckMySum
## A simple checksum viewer made in Python!

### Description:

###### This is a simple program made in Python that's capable of computing some of the most common hash algorithms available for the Python's 'hashlib' library.

###### The program will require you to choose an algorithm (md5, sha1, sha224, sha256, sha384, sha512) and the files that you'll want to check. Here are the possible options:

    Usage: CheckMySum.exe [-a/--algorithm] ALGORITHM [-f/--files] FILES ... [-u/--upper] [-g/--generate GENERATE]
    -------------------------------------------------------------------------------------------------------------

    '-h', '--help'          Show the help message and exit
    '-v', '--version'       Show the programs version and exit
    
    '-a', '--algorithm'     Defines the hash algorithm to be used
    '-f', '--files'         Select the files to be checked
    '-u', '--upper'         Display the hash value in uppercase letters
    '-g', '--generate'      Generate a checksum file for the files checked

###### You can generate a hash file with the results computed by the program, just use the '-g' or '--generate' option and input a filename for the hash file to be created.

###### You can display the hash in uppercase letters by using the '-u' or '--upper' option. By default the hash will be displayed in downcase letters.

###### I've made this program to be used on Windows but it can run just fine on other operating systems (like GNU/Linux, BSD and OSX), as long as you have Python 2. 

## [Check the CHANGELOG.txt for more informations!](https://raw.github.com/Wolfterro/CheckMySum/master/CHANGELOG.txt)

### Requirements:

#### Compiling:
- Python 2.7
- PyInstaller
- Microsoft Visual C++ 2010 Redistributable (Windows only)

##### To compile the program on Windows, just execute the file 'build.bat'.
##### To compile the program on Linux, just execute the file 'build.sh'.
##### **PS**: PyInstaller must be recognized as an internal command on Windows CMD or in the shell if you're using GNU/Linux or other UNIX-like operating system. 

#### Executing directly:

##### You can execute the program directly if you don't want to compile it or in case you don't have PyInstaller:

    python2 src/Main.py [-a/--algorithm] ALGORITHM [-f/--files] FILES ... [-u/--upper] [-g/--generate GENERATE]

### Download:

##### **PS**: Binary files doesn't require Python to be installed!
#### ***Binary (Windows):*** https://github.com/Wolfterro/CheckMySum/releases/tag/v1.0.1-Windows
#### ***Source code:*** https://github.com/Wolfterro/CheckMySum/archive/master.zip


CHANGELOG.txt
CheckMySum
==========

A simple checksum viewer made in Python! 
----------------------------------------

CHANGELOG:
==========

VERSION 1.0.1:
- Uppercase option added.
- Generate hash file option added.

# In progress:
- Better management of the CheckMySum class structure for other usages (e.g. GUI, standalone module, etc).

#---------------------------------------------------------------------------------------------------------------

VERSION 1.0.0:
- Initial version of the program.

#---------------------------------------------------------------------------------------------------------------

src/CheckMySum.py
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.0.0 - Python 2.x
# Data: 30/03/2017
#===================================

# Imports gerais
# ==============
from __future__ import print_function

import os
import sys
import hashlib

# Imports do programa
# ===================
from GlobalVars import GlobalVars

# Definindo a codificação padrão para UTF-8.
# ==========================================
reload(sys)
sys.setdefaultencoding('utf-8')

# Classe Principal do programa
# ============================
class CheckMySum(object):
	# Inicializando a classe com o tipo de algoritmo e nome do arquivo
	# ================================================================
	def __init__(self, Algorithm, Filename):
		self.Algorithm = Algorithm
		self.Filename = Filename
		self.Hash = ""
		self.ProcessFile()

	# Iniciando processo de checksum do arquivo selecionado
	# =====================================================
	def ProcessFile(self):
		if self.CheckFileAvailability(self.Filename):
			self.GetCheckSum()
			self.PrintCheckSum()

	# Verificando a existência do arquivo selecionado
	# ===============================================
	def CheckFileAvailability(self, Filename):
		if not os.path.isfile(os.path.abspath(self.Filename)):
			print("Error! File '%s' not available! Skipping..." % (self.Filename))
			return False

		return True

	# Resgatando o checksum do arquivo selecionado
	# ============================================
	def GetCheckSum(self):
		if self.Algorithm == "md5":
			Algo = hashlib.md5()
		elif self.Algorithm == "sha1":
			Algo = hashlib.sha1()
		elif self.Algorithm == "sha224":
			Algo = hashlib.sha224()
		elif self.Algorithm == "sha256":
			Algo = hashlib.sha256()
		elif self.Algorithm == "sha384":
			Algo = hashlib.sha384()
		elif self.Algorithm == "sha512":
			Algo = hashlib.sha512()
		else:
			Algo = hashlib.md5()

		print("Checking...", end="")
		try:
			with open(os.path.abspath(self.Filename), "rb") as file:
				for f in file:
					Algo.update(f)
		
			self.Hash = Algo.hexdigest()
		except Exception as e:
			self.Hash = "N/A - %s" % (e)

	# Imprimindo o checksum do arquivo escolhido
	# ==========================================
	def PrintCheckSum(self):
		print("\r%s - %s" % (self.Filename, self.Hash))

src/old/v1.0.0/GlobalVars.py
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.0.0 - Python 2.x
# Data: 30/03/2017
#===================================

# Classe de variáveis globais
# ===========================
class GlobalVars:
	Version = "1.0.0"

src/Main.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-

'''
The MIT License (MIT)

Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
'''

#===================================
# Criado por: Wolfterro
# Versão: 1.0.0 - Python 2.x
# Data: 30/03/2017
#===================================

# Imports gerais
# ==============
from __future__ import print_function
import sys
import hashlib
import argparse

# Imports do programa
# ===================
from CheckMySum import CheckMySum
from GlobalVars import GlobalVars

# Definindo a codificação padrão para UTF-8.
# ==========================================
reload(sys)
sys.setdefaultencoding('utf-8')

# Método de entrada do programa
# =============================
def main():
	if ("-v" in sys.argv or "--version" in sys.argv) and len(sys.argv) <= 2:
		print("==========================")
		print("CheckMySum - Version %s" % (GlobalVars.Version))
		print("==========================\n")
		
		print(" *** This program is licensed under the MIT License ***\n")
		
		print("Copyright (c) 2017 Wolfgang Almeida <wolfgang.almeida@yahoo.com>")
		print("GitHub's repository: %s\n" % ("https://github.com/Wolfterro/CheckMySum"))
		sys.exit(0)

	parser = argparse.ArgumentParser()
	parser.add_argument("-a", "--algorithm", choices=hashlib.algorithms_guaranteed, help="Defines the hash algorithm to be used", required=True)
	parser.add_argument("-f", "--files", nargs="+", help="Select the files to be checked", required=True)
	parser.add_argument("-v", "--version", action="store_true", help="Show the programs version and exit", required=False)

	args = parser.parse_args()

	HashAlgorithm = args.algorithm.lower()
	for file in args.files:
		MySum = CheckMySum(HashAlgorithm, file)

# Inicializando programa
# ======================
if __name__ == "__main__":
	main()